### ISSUE TITLE
[mac] codesign option doesn't work on macos 11.6.6

### ISSUE THREAD

**Issue by umjammer** [fvarrui/JavaPackager#267]
I'm submitting a…

- [x] bug report
- [ ] feature request
- [ ] other

**Short description of the issue/suggestion:**

enabling codesign on macos  11.6.6 cause error

**Steps to reproduce the issue/enhancement:**

1. pom.xml like
```xml
<bundleJre>true</bundleJre>
<macConfig>
  <codesignApp>true</codesignApp>
```
2.`$ mvn package`

**What is the expected behavior?**

package will be created

**What is the current behavior?**

got am error

**Do you have outputs, screenshots, demos or samples which demonstrate the problem or enhancement?**

```
[ERROR] Failed to execute goal io.github.fvarrui:javapackager:1.7.0-SNAPSHOT:package (default) on project vavi-apps-comicviewer: Command execution failed: codesign --force --options runtime --deep --entitlements /Volumes/nsanomac4_ssd1/Users/nsano/src/vavi/vavi-apps-comicviewer/target/assets/entitlements.plist --sign - /Volumes/nsanomac4_ssd1/Users/nsano/src/vavi/vavi-apps-comicviewer/target/zzzViewer/zzzViewer.app -> [Help 1]
```

**What is the motivation / use case for changing the behavior?**

```diff
diff --git a/src/main/java/io/github/fvarrui/javapackager/packagers/MacPackager.java b/src/main/java/io/github/fvarrui/javapackager/packagers/MacPackager.java
index e7da3a5..bfad208 100644
--- a/src/main/java/io/github/fvarrui/javapackager/packagers/MacPackager.java
+++ b/src/main/java/io/github/fvarrui/javapackager/packagers/MacPackager.java
@@ -234,7 +234,7 @@ public class MacPackager extends Packager {
                codesignArgs.add(appFile);
                
                // run codesign
-               CommandUtils.execute("codesign", codesignArgs.toArray(new Object[codesignArgs.size()]));
+               CommandUtils.execute("sudo", "codesign", codesignArgs.toArray(new Object[codesignArgs.size()]));
        }
 
 }
```

add `sudo`, it works.

**Please tell us about your environment:**

- JavaPackager version:
1.7.0-SNAPSHOT
```
$ git log                                                                                                                                                                                                                         devel
commit 50e56362865bfbb6429bbb593258bd3d0c3d7c66 (HEAD -> devel)
```
- OS version:
MacOS Big Sur 11.6.6
```
Darwin nsanomac4 20.6.0 Darwin Kernel Version 20.6.0: Tue Apr 19 21:04:45 PDT 2022; root:xnu-7195.141.29~1/RELEASE_X86_64 x86_64
```
- JDK version:
```
$ java -version                                                                                                                                                                                                  main
openjdk version "19" 2022-09-20
OpenJDK Runtime Environment Homebrew (build 19)
OpenJDK 64-Bit Server VM Homebrew (build 19, mixed mode, sharing)
```
- Build tool: 
	- [x] Maven 
	- [ ] Gradle 

**Other information** (e.g. related issues, suggestions how to fix, links for us to have context)

after patch above, codesigning wroks fine, but got a new error.

---

```
[ERROR] Failed to execute goal io.github.fvarrui:javapackager:1.7.0-SNAPSHOT:package (default) on project vavi-apps-comicviewer: Could not remove folder /Volumes/nsanomac4_ssd1/Users/nsano/src/vavi/vavi-apps-comicviewer/target/zzzViewer/zzzViewer.app/Contents/PlugIns/jre/Contents/Home/legal: Unable to delete directory /Volumes/nsanomac4_ssd1/Users/nsano/src/vavi/vavi-apps-comicviewer/target/zzzViewer/zzzViewer.app/Contents/PlugIns/jre/Contents/Home/legal. -> [Help 1]
```

this error says that "could not remove the directory because it's not empty".

```
ls -la /Volumes/nsanomac4_ssd1/Users/nsano/src/vavi/vavi-apps-comicviewer/target/zzzViewer/zzzViewer.app/Contents/PlugIns/jre/Contents/Home/legal   
total 0
drwxr-xr-x 44 nsano staff 1408 2022-10-19 19:01 .
drwxr-xr-x 11 nsano staff  352 2022-10-19 19:01 ..
drwxr-xr-x  5 nsano staff  160 2022-10-19 19:01 java.datatransfer
drwxr-xr-x  5 nsano staff  160 2022-10-19 19:01 java.desktop
drwxr-xr-x  5 nsano staff  160 2022-10-19 19:01 java.naming
drwxr-xr-x  5 nsano staff  160 2022-10-19 19:01 java.net.http
drwxr-xr-x  5 nsano staff  160 2022-10-19 19:01 java.rmi
drwxr-xr-x  5 nsano staff  160 2022-10-19 19:01 java.scripting
drwxr-xr-x  5 nsano staff  160 2022-10-19 19:01 java.smartcardio
drwxr-xr-x  5 nsano staff  160 2022-10-19 19:01 java.sql
:
```

so i did 2nd patch like

```diff
diff --git a/src/main/java/io/github/fvarrui/javapackager/utils/FileUtils.java b/src/main/java/io/github/fvarrui/javapackager/utils/FileUtils.java
index d2f170f..7443963 100644
--- a/src/main/java/io/github/fvarrui/javapackager/utils/FileUtils.java
+++ b/src/main/java/io/github/fvarrui/javapackager/utils/FileUtils.java
@@ -4,6 +4,7 @@ import static org.apache.commons.io.FileUtils.copyDirectoryToDirectory;
 import static org.apache.commons.io.FileUtils.copyFile;
 import static org.apache.commons.io.FileUtils.copyInputStreamToFile;
 import static org.apache.commons.io.FileUtils.deleteDirectory;
+import static org.apache.commons.io.FileUtils.forceDelete;
 import static org.apache.commons.io.FileUtils.moveDirectoryToDirectory;
 import static org.apache.commons.io.FileUtils.moveFileToDirectory;
 import static org.apache.commons.io.FileUtils.readFileToString;
@@ -14,16 +15,20 @@ import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.UncheckedIOException;
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.nio.file.StandardCopyOption;
 import java.util.Arrays;
+import java.util.Comparator;
 import java.util.List;
 import java.util.function.Function;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import org.apache.commons.io.IOUtils;
 import org.apache.commons.lang3.StringUtils;
@@ -243,8 +248,10 @@ public class FileUtils {
        
        public static void removeFolder(File folder) throws Exception {
                Logger.info("Removing folder [" + folder + "]");                
-               try {
-                       deleteDirectory(folder);
+               try (Stream<Path> s = Files.walk(Paths.get(folder.getPath()))) {
+                       s.sorted(Comparator.reverseOrder())
+                                       .map(Path::toFile)
+                                       .forEach(File::delete);
                } catch (IOException e) {
             throw new Exception("Could not remove folder " + folder, e);
                }
```
---

after 2 patches, codesign works well on my mac env.
but 1st patch is not good because `sudo` is hardcoded. it's better to outsource about codesign command settings.
please fix about those in your way.

**Comment by fvarrui** [comment]
Hi @umjammer!
Thanks for proposing patches.

I don't know why [`org.apache.commons.io.FileUtils.deleteDirectory`](https://commons.apache.org/proper/commons-io/javadocs/api-2.6/org/apache/commons/io/FileUtils.html#deleteDirectory-java.io.File-) is not working fine for you, so far it has been doing its job well 🤷‍♂️ . It's a bit weird.

Maybe those problems are related to your permissions. Why do you need to use `sudo`?? 😨 

PS: `legal` folder is removed because `codesign` fails.

**Comment by kerner1000** [comment]
I can confirm there is a bug in [org.apache.commons.io.FileUtils.deleteDirectory](https://commons.apache.org/proper/commons-io/javadocs/api-2.6/org/apache/commons/io/FileUtils.html#deleteDirectory-java.io.File-). It is fixed in the latest version.
Just update the dependency:

	implementation 'commons-io:commons-io:2.7'
to
	implementation 'commons-io:commons-io:2.11.0'

**Comment by fvarrui** [comment]
Ok, I'll try to upgrade `commons-io` in v1.7.0.
Thanks!

**Comment by fvarrui** [comment]
**commons-io** upgraded to 2.11.0 in branch `issue-239` ... Please, try the fix and give some feedback. Thanks!

**Comment by dipesh47** [comment]
Same issue with macos 13.1 and fvarrui verson 1.6.7.

**Comment by fvarrui** [comment]
> Same issue with macos 13.1 and fvarrui verson 1.6.7.

Have you tested it using version 1.7.0-SNAPSHOT?

```bash
git clone https://github.com/fvarrui/JavaPackager --branch devel
cd JavaPackager
./gradlew publishToMavenLocal
```

> Use JDK 8-11 to compile

**Comment by dipesh47** [comment]
Same issue with 1.7.0, i have not used SNAPSHOT version. Is it not fixed in 1.7.0?

**Comment by fvarrui** [comment]
> Same issue with 1.7.0, i have not used SNAPSHOT version. Is it not fixed in 1.7.0?

1.7.0 is not released yet, just SNAPSHOT

**Comment by fvarrui** [comment]
JavaPackager 1.7.0 released to Maven Central
